core/src/letter_distribution.c

```diff
@@ -150,50 +153,148 @@ uint8_t human_readable_letter_to_machine_letter(
+//  * 0 for subsequent bytes in the code point or
+//  * -1 for invalid UTF8 bytes
+int get_number_of_utf8_bytes_for_code_point(uint8_t byte) {
+  int number_of_bytes = -1;
+  if ((byte & 0xC0) == 0x80) {
+    // Subsequent byte in a code point
+    return 0;
+  } else if ((byte & 0x80) == 0x00) {
+    // Single-byte UTF-8 character
+    return 1;
+  } else if ((byte & 0xE0) == 0xC0) {
+    // Two-byte UTF-8 character
+    return 2;
+  } else if ((byte & 0xF0) == 0xE0) {
+    // Three-byte UTF-8 character
+    return 3;
+  } else if ((byte & 0xF8) == 0xF0) {
+    // Four-byte UTF-8 character
+    return 4;
+  }
+  return number_of_bytes;
+}
+
 // Convert a string of arbitrary characters into an array of machine letters,
 // returning the number of machine letters. This function does not allocate
 // the ml array; it is the caller's responsibility to make this array big
-// enough.
+// enough. This function will return -1 if it encounters an invalid letter.
 // Note: This is a slow function that should not be used in any hot loops.
 int str_to_machine_letters(const LetterDistribution *letter_distribution,
                            const char *str, bool allow_played_through_marker,
-                           uint8_t *mls) {
+                           uint8_t *mls, size_t mls_size) {
 
   int num_mls = 0;
-  int num_bytes = string_length(str);
-  int i = 0;
-  int prev_i = -1;
-  while (i < num_bytes) {
-    for (int j = i + letter_distribution->max_tile_length; j > i; j--) {
-      if (j > num_bytes) {
-        continue;
+  size_t num_bytes = string_length(str);
+  // Use +1 for the null terminator
+  char current_letter[MAX_LETTER_CHAR_LENGTH + 1];
+  int current_letter_byte_index = 0;
+  bool building_multichar_letter = false;
+  int current_unicode_bytes_remaining = 0;
+  int number_of_chars_in_builder = 0;
+
+  // While writing to mls, this loop verifies the following:
+  // - absence of nested multichar characters
+  // - bijection between the set of start and end multichar delimiters
+  // - multichar characters are nonempty
+  for (size_t i = 0; i < num_bytes; i++) {
+    char current_char = str[i];
+    switch (current_char) {
+    case MULTICHAR_START_DELIMITIER:
+      if (building_multichar_letter || current_unicode_bytes_remaining > 0) {
+        return -1;
+      }
+      building_multichar_letter = true;
+      break;
+    case MULTICHAR_END_DELIMITIER:
+      // Return an error if
+      // - multichar is building
+      // - multichar has fewer than two bytes
+      // - unicode code point is being built
+      if (!building_multichar_letter || number_of_chars_in_builder < 2 ||
+          current_unicode_bytes_remaining > 0) {
+        return -1;
+      }
+      building_multichar_letter = false;
+      break;
+    default:
+      if (current_letter_byte_index == MAX_LETTER_CHAR_LENGTH) {
+        // Exceeded max char length
+        return -1;
+      }
+
+      int bytes_remaining =
```

Comment:
bytes_remaining may need a better name (i keep reading it as the bytes remaining for the current codepoint)